File: 53-maximum-subarray/maximum-subarray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxSubArray = function(nums) {
    let maxSum = nums[0], prevSum = 0;
    for (let n of nums) {
        prevSum = Math.max(prevSum + n, n);
        maxSum = Math.max(prevSum, maxSum);
    }
    return maxSum;
};

File: 53-maximum-subarray/maximum-subarray.swift
class Solution {
    func maxSubArray(_ nums: [Int]) -> Int {
        var dp: [Int] = nums.map{$0}
        var maxSum: Int = dp[0]
        for i in 1..<dp.count {
            dp[i] = max(dp[i], dp[i] + dp[i-1])
            maxSum = max(maxSum, dp[i])
        }
        return maxSum
    }
}